overlord/ifacestate/ifacestate_test.go

```diff
@@ -2373,6 +2377,221 @@ func (s *interfaceManagerSuite) TestDoSetupSnapSecurityKeepsExistingConnectionSt
+	s.state.Lock()
+	s.state.Set("conns", map[string]interface{}{
+		"consumer:plug producer:slot": map[string]interface{}{
+			"interface":   "test",
+			"plug-static": map[string]interface{}{"attr": "old-plug-attr"},
+			"slot-static": map[string]interface{}{"attr": "old-slot-attr"},
+		},
+	})
+	s.state.Unlock()
+
+	// Add consumer and producer snaps, with a plug and slot respectively, each
+	// carrying a single attribute with a "new" value. The "new" value is in
+	// contrast to the old value in the connection state.
+	const consumerYaml = `
+name: consumer
+version: 1
+plugs:
+ plug:
+  interface: test
+  attr: new-plug-attr
+`
+	const producerYaml = `
+name: producer
+version: 1
+slots:
+ slot:
+  interface: test
+  attr: new-slot-attr
+`
+	s.mockSnap(c, producerYaml)
+	s.mockSnap(c, consumerYaml)
+
+	// Create a connection reference, it's just verbose and used a few times
+	// below so it's put up here.
+	connRef := &interfaces.ConnRef{
+		PlugRef: interfaces.PlugRef{Snap: "consumer", Name: "plug"},
+		SlotRef: interfaces.SlotRef{Snap: "producer", Name: "slot"}}
+
+	// Add a test security backend and a test interface. We want to use them to
+	// observe the interaction with the security backend and to allow the
+	// interface manager to keep the test plug and slot of the consumer and
+	// producer snaps we introduce below.
+	secBackend := &ifacetest.TestSecurityBackend{
+		BackendName: "test",
+		SetupCallback: func(snapInfo *snap.Info, opts interfaces.ConfinementOptions, repo *interfaces.Repository) error {
+			// Whenever this function is invoked to setup security for a snap
+			// we check the connection attributes that it would act upon.
+			// Because of how connection state is refreshed we never expect to
+			// see the old attribute values.
+			conn, err := repo.Connection(connRef)
+			c.Assert(err, IsNil)
+			c.Check(conn.Plug.StaticAttrs(), DeepEquals, map[string]interface{}{"attr": "new-plug-attr"})
+			c.Check(conn.Slot.StaticAttrs(), DeepEquals, map[string]interface{}{"attr": "new-slot-attr"})
+			return nil
+		},
+	}
+	s.mockSecBackend(c, secBackend)
+	s.mockIfaces(c, &ifacetest.TestInterface{InterfaceName: "test"})
+
+	// Create the interface manager. This indirectly adds the snaps to the
+	// repository and re-connects them using the stored connection information.
+	mgr := s.manager(c)
+
+	// Inspect the repository connection data. The data no longer refers to the
+	// old connection attributes because they were updated when the connections
+	// were reloaded from the state.
+	repo := mgr.Repository()
+	conn, err := repo.Connection(connRef)
+	c.Assert(err, IsNil)
+	c.Check(conn.Plug.StaticAttrs(), DeepEquals, map[string]interface{}{"attr": "new-plug-attr"})
+	c.Check(conn.Slot.StaticAttrs(), DeepEquals, map[string]interface{}{"attr": "new-slot-attr"})
+
+	// Because of the fact that during testing the system key always
+	// mismatches, the security setup is performed.
+	c.Check(secBackend.SetupCalls, HasLen, 2)
+}
```

Comment:
I know backend setup callback check sort of covers it, but for completeness I'd check get "conns" from state here and check the expected map. That's also probably the easiest way to ensure the "canary" connection I mentioned above is in place.

Reply:
+1